evm/src/executor/inspector/cheatcodes/expect.rs

```diff
@@ -26,71 +29,152 @@ static DUMMY_CREATE_ADDRESS: Address =
+    expected_revert_address: H160,
+    reverts_by_address: BTreeMap<H160, HashSet<Bytes>>,
+) -> Result<(Option<Address>, Bytes)> {
+    let address_reverts = reverts_by_address.get(&expected_revert_address);
+
+    if address_reverts.is_none() {
+        return Err(fmt_err!(
+            "The expected revert address {:#?} did not revert",
+            expected_revert_address,
+        ))
     }
 
+    #[allow(clippy::mutable_key_type)]
+    let address_reverts = address_reverts.unwrap();
+
     // If None, accept any revert
     let expected_revert = match expected_revert {
         Some(x) => x,
-        None => return success_return!(),
+        None => return success_return!(is_create),
     };
 
-    if !expected_revert.is_empty() && retdata.is_empty() {
-        bail!("Call reverted as expected, but without data");
+    let address_reverts = address_reverts
+        .iter()
+        .map(|revert_data| actual_error_from_bytes(revert_data.clone()).to_string())
+        .collect::<HashSet<_>>();
+
+    if address_reverts.contains(&expected_revert.to_string()) {
+        success_return!(is_create)
     }
+    // Empty revert data
+    else if address_reverts.contains("0x") {
+        bail!(
+            "The expected revert address {:#?} reverted as expected, but without data",
+            expected_revert_address
+        );
+    } else {
+        Err(fmt_err!(
+            "The expected revert address {:#?} did not revert with the expected revert data: {}",
+            expected_revert_address,
+            stringify(expected_revert),
+        ))
+    }
+}
 
-    let mut actual_revert = retdata;
-    if actual_revert.len() >= 4 &&
-        matches!(actual_revert[..4].try_into(), Ok(ERROR_PREFIX | REVERT_PREFIX))
-    {
-        if let Ok(bytes) = Bytes::decode(&actual_revert[4..]) {
-            actual_revert = bytes;
-        }
+/// Verifies that the current call is reverting and that it reverted with `expected_revert`
+fn handle_expect_revert_with_data(
+    is_create: bool,
+    expected_revert: Option<&Bytes>,
+    status: InstructionResult,
+    retdata: Bytes,
+) -> Result<(Option<Address>, Bytes)> {
+    ensure!(!matches!(status, return_ok!()), "Call did not revert as expected");
+
+    // If None, accept any revert
+    let expected_revert = match expected_revert {
+        Some(x) => x,
+        None => return success_return!(is_create),
+    };
+
+    if !expected_revert.is_empty() && retdata.is_empty() {
+        bail!("Call reverted as expected, but without data");
     }
 
+    let actual_revert = actual_error_from_bytes(retdata);
     if actual_revert == *expected_revert {
-        success_return!()
+        success_return!(is_create)
     } else {
-        let stringify = |data: &[u8]| {
-            String::decode(data)
-                .ok()
-                .or_else(|| std::str::from_utf8(data).ok().map(ToOwned::to_owned))
-                .unwrap_or_else(|| format!("0x{}", hex::encode(data)))
-        };
```

Comment:
ah right, the stringify function is this extracted to an outside func?

Reply:
yup, previously it was a closure defined inside the `handle_expect_revert` so I exported it into its own thing and reused it in both `expect_revert_...` functions.